src/stan/model/indexing/assign_varmat.hpp

```diff
@@ -55,16 +131,19 @@ namespace model {
  * @throw std::out_of_range If the index is out of bounds.
  */
 template <typename VarVec, typename U, require_var_vector_t<VarVec>* = nullptr,
-          require_var_t<U>* = nullptr,
-          require_floating_point_t<value_type_t<U>>* = nullptr>
+          require_stan_scalar_t<U>* = nullptr>
 inline void assign(VarVec&& x, const U& y, const char* name, index_uni idx) {
   stan::math::check_range("var_vector[uni] assign", name, x.size(), idx.n_);
   const auto coeff_idx = idx.n_ - 1;
   double prev_val = x.val().coeffRef(coeff_idx);
-  x.vi_->val_.coeffRef(coeff_idx) = y.val();
+  x.vi_->val_.coeffRef(coeff_idx) = stan::math::value_of(y);
   stan::math::reverse_pass_callback([x, y, coeff_idx, prev_val]() mutable {
     x.vi_->val_.coeffRef(coeff_idx) = prev_val;
-    y.adj() += x.adj().coeffRef(coeff_idx);
+    using stan::math::forward_as;
+    using stan::math::var;
+    if (!is_constant<U>::value) {
+      forward_as<var>(y).adj() += x.adj().coeffRef(coeff_idx);
```

Comment:
```suggestion
      math::adjoint_of(y) += x.adj().coeffRef(coeff_idx);
```
This can be simplified.

Reply:
Nice! Is there a reason we don't have this for the `matrix<var>` case?

Reply:
It was probably implemented before `adjoint_of`.